artifact/v3/header/header_info.cpp

```diff
@@ -0,0 +1,195 @@
+	for (size_t i = 0; i < vector_size; ++i) {
+		auto expected_element =
+			j.Get(i).and_then([](const json::Json &j) { return j.Get("type"); });
+		if (!expected_element) {
+			return expected::unexpected(parser_error::MakeError(
+				parser_error::Code::ParseError,
+				"Failed to get the type from the payload" + expected_element.error().message));
+		}
+		auto json_element = expected_element.value();
+		if (json_element.IsString()) {
+			string payload_type = json_element.GetString().value();
+			if (payload_type == "") {
+				return expected::unexpected(
+					parser_error::MakeError(parser_error::Code::ParseError, "Empty Payload type"));
+			}
+			Payload type {Payload::RootfsImage};
+			if (payload_type != "rootfs-image") {
+				type = Payload::ModuleImage;
+			}
+			vector_elements.push_back({type, payload_type});
+		} else if (json_element.IsNull()) {
+			vector_elements.push_back({Payload::EmptyPayload, ""});
+		} else {
+			return expected::unexpected(
+				parser_error::MakeError(parser_error::Code::ParseError, "Unexpected payload type"));
+		}
+	}
+	return vector_elements;
+}
+
+ExpectedHeaderInfo Parse(io::Reader &reader) {
+	log::Trace("Parse(header-info)");
+
+	Info info {};
+
+	auto expected_json = json::Load(reader);
+
+	if (!expected_json) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError,
+			"Failed to parse the header JSON: " + expected_json.error().message));
+	}
+
+	const json::Json header_info_json = expected_json.value();
+
+	//
+	// Payloads (required)
+	//
+	log::Trace("Parsing the payloads");
+
+	auto payloads = header_info_json.Get("payloads").and_then(ToPayloadTypes);
+	if (!payloads) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError,
+			"Failed to parse the header-info payloads JSON: " + payloads.error().message));
+	}
+	info.payloads = payloads.value();
+
+	//
+	// provides (required)
+	//
+	log::Trace("Parsing the header-info artifact_provides");
+
+	auto provides = header_info_json.Get("artifact_provides");
+	if (!provides) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError,
+			"Failed to parse the header-info artifact_provides JSON: " + provides.error().message));
+	}
+
+	// provides:artifact_name (required)
+	log::Trace("Parsing the header-info provides:artifact_name");
+	auto artifact_name = provides.value().Get("artifact_name").and_then(json::ToString);
+
+	if (!artifact_name) {
+		return expected::unexpected(
+			parser_error::MakeError(parser_error::Code::ParseError, artifact_name.error().message));
+	}
+	info.provides.artifact_name = artifact_name.value();
+
+	// provides:artifact_group (optional)
+	log::Trace("Parsing the header-info provides:artifact_group (if any)");
+	auto artifact_group = provides.value().Get("artifact_group").and_then(json::ToString);
+	if (!artifact_group && artifact_group.error().code.value() != json::KeyError) {
```

Comment:
I'd do `Get("artifact_group")` and then `GetString()` if it returns something, but YMMV.

Reply:
Yeah, could've. I think they both make sense here, so staying with the current impl